pkg/webhooks/podimagespec/podimagespec.go

```diff
@@ -0,0 +1,283 @@
+	rules         = []admissionregv1.RuleWithOperations{
+		{
+			Operations: []admissionregv1.OperationType{
+				admissionregv1.Create,
+			},
+			Rule: admissionregv1.Rule{
+				APIGroups:   []string{""},
+				APIVersions: []string{"v1"},
+				Resources:   []string{"pods"},
+				Scope:       &scope,
+			},
+		},
+	}
+	log        = logf.Log.WithName(WebhookName)
+	imageRegex = regexp.MustCompile(`(image-registry\.openshift-image-registry\.svc:5000\/)(?P<namespace>\S*)(/)(?P<image>\S*)(:)(?P<tag>\S*)`)
+)
+
+// PodImageSpecWebhook mutates an image spec in a pod
+type PodImageSpecWebhook struct {
+	s          *runtime.Scheme
+	configV1   configv1.Config
+	kubeClient client.Client
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *PodImageSpecWebhook {
+	scheme := runtime.NewScheme()
+	err := configv1.Install(scheme)
+	if err != nil {
+		panic(err)
+	}
+	return &PodImageSpecWebhook{
+		s: scheme,
+	}
+}
+
+// CheckImageRegistryStatus checks the status of the image registry service
+func (s *PodImageSpecWebhook) CheckImageRegistryStatus(ctx context.Context) (bool, error) {
+	var err error
+	if s.kubeClient == nil {
+		s.kubeClient, err = k8sutil.KubeClient(s.s)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	err = s.kubeClient.Get(ctx, client.ObjectKey{Name: "cluster"}, &s.configV1)
+	if err != nil {
+		return false, fmt.Errorf("failed to get image registry config: %v", err)
+	}
+
+	if s.configV1.Spec.ManagementState == operatorv1.Managed {
+		return true, nil
+	} else if s.configV1.Spec.ManagementState == operatorv1.Removed {
+		return false, nil
+	}
+
+	return false, fmt.Errorf("unknown managementState: %s", s.configV1.Spec.ManagementState)
+}
+
+// Authorized implements Webhook interface
+func (s *PodImageSpecWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	//Implement authorized next
+	return s.authorizeOrMutate(request)
+}
+
+// order of operations
+// check regex
+// pull image-registry status
+// pull image spec
+
+// kube client?
+// generate on boot
+// permissions for webhook service account
+// get kubeconfig from webhook service acount
+// make sure service account token is in the pod
+// ask micheal shen about where to put kubeclient
+
+// run performance tests
+
+// authorizeOrMutate decides whether the Request requires mutation before it's allowed to proceed.
+func (s *PodImageSpecWebhook) authorizeOrMutate(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+	pod, err := s.renderPod(request)
+	if err != nil {
+		log.Error(err, "Couldn't render a Pod from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+	//1. Regex match
+	mutated := false
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		imageMutated, err := mutateContainerImageSpec(&container.Image)
+		if err != nil {
+			return admissionctl.Errored(http.StatusBadRequest, err)
+		}
+		if imageMutated {
+			mutated = true
+		}
+	}
```

Comment:
This isn't just checking for the regex, this is actually mutating the images.. but you can't mutate the images without doing a lookup as to what you're mutating them to do which is really expensive.

This needs to be a smaller check.. just check IF the pods need to be mutated, but don't actually do it yet. The idea is to offramp the traffic as quickly as possible.